pkg/rbd/rbd.go

```diff
@@ -27,10 +27,89 @@ import (
 	"k8s.io/utils/exec"
 )
 
-// PluginFolder defines the location of rbdplugin
+/*
+RADOS omaps usage:
+
+This note details how we preserve idempotent nature of create requests and retain the relationship
+between orchestrator (CO) generated Names and plugin generated names for images and snapshots
+
+The implementation uses Ceph RADOS omaps to preserve the relationship between request name and
+generated image (or snapshot) name. There are 4 types of omaps in use,
+- A "csi.volumes.directory.[csi-id]" (or volOMap+.+CsiInstanceID), we call this the volOMap
+  - stores keys named using the CO generated names for volume requests
+  - keys are named "csi.volume."+[CO generated VolName]
+  - Key value contains the RBD image name that is created or will be created, for the CO provided name
+
+- A "csi.snaps.directory.[csi-id]" (or snapsOMap+.+CsiInstanceID), we refer to this as the snapsOMap
+  - stores keys named using the CO generated names for snapshot requests
+  - keys are named "csi.snap."+[CO generated SnapName]
+  - Key value contains the RBD snapshot name that is created or will be created, for the CO provided name
+
+- A per image omap named "csi.volume."+[RBD image name], we refer to this as the imgOMap
+  - stores a single key named "csi.volname", that has the value of the CO generated VolName that
+  this image refers to
+
+- A per snapshot omap named "csi.snap."+[RBD snapshot name], we refer to this as the snapOMap
+  - stores a key named "csi.snapname", that has the value of the CO generated SnapName that this
+  snapshot refers to
+  - also stores another key named "csi.source", that has the value of the image name that is the
+  source of the snapshot
+
+Creation of omaps:
+When a volume create request is received (or a snapshot create, the snapshot is not detailed in this
+	comment further as the process is similar),
+- The volOMap is consulted to find if there is already a key with the CO VolName, and if present,
+it is used to read its references to reach the RBD image that backs this VolName, to check if the
+RBD image can satisfy the requirements for the request
+  - If during the process of checking the same, it is found that some linking information is stale
+  or missing, the corresponding keys upto the key in the volOMap is cleaned up, to start afresh
+- If the key is not found, or was cleaned up, the request is treated as a fresh create request, and
```

Comment:
I am unclear which part this review comment refers to, but assuming the following and responding.

The comment is to do the following,
- `rados [...] create csi.volume.<image-uuid>`
  - Check for EEXISTS, to deal with image-uuid in use, and handle failure paths
- On success,
  - `rados [...] setomapval csi.volume.<image-uuid> csi.volname <volume-name>`

Instead of the current method of using just setomapval, to create the omap and set the key and value, in one shot.

-----------------------------

Ack, just tried using setomapval on an existing omap and key, and as expected it overwrites the value in the key, which is not what we require. Creating the omap first (and dealing with exists situations) and then setting the key is the safe thing to do.

Will address in an update to this PR.